sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
-	},
-	{
-		.name = "SDW0-Capture",
-		.id = 1,
-		.no_pcm = 1,
-		.dpcm_capture = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw0_pin3, sdw0_codec, platform),
-	},
-	{
-		.name = "SDW1-Playback",
-		.id = 2,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw1_pin2, sdw1_codec, platform),
-	},
-	{
-		.name = "SDW3-Capture",
-		.id = 4,
-		.no_pcm = 1,
-		.dpcm_capture = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw3_pin2, sdw3_codec, platform),
-	},
+static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+{
+	const struct snd_soc_acpi_link_adr *link;
+	int num = 0;
+
+	link = mach_params->links;
+	while (link->num_adr) {
+		int part_id;
+
+		part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+		/* playback & capture */
+		if (part_id == 0x711)
+			num++;
+
+		num++;
+		link++;
+	}
+
+	return num;
+}
+
+static void init_dai_link(struct snd_soc_dai_link *dai_links, int id,
+			  char *name,
+			  struct snd_soc_dai_link_component *cpus,
+			  char *cpu_name,
+			  struct snd_soc_dai_link_component *codecs,
+			  int codecs_num)
+{
+	dai_links[id].id = id;
+	dai_links[id].name = name;
+	dai_links[id].platforms = platform_component;
+	dai_links[id].num_platforms = ARRAY_SIZE(platform_component);
+	dai_links[id].nonatomic = true;
+	dai_links[id].no_pcm = 1;
+	dai_links[id].cpus = &cpus[id];
+	dai_links[id].num_cpus = 1;
+	dai_links[id].cpus->dai_name = cpu_name;
+	dai_links[id].codecs = codecs;
+	dai_links[id].num_codecs = codecs_num;
+}
+
+static int create_sdw_codec_dai(struct device *dev,
+				struct snd_soc_dai_link *dai_links, int id,
+				struct snd_soc_dai_link_component *cpus,
+				const struct snd_soc_acpi_link_adr *link)
+{
+	unsigned int part_id, unique_id, mfg_id, link_id, class_id;
+	struct snd_soc_dai_link_component *codec;
+	int i, j, dai_num = 1;
+
+	codec = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component) *
+			     link->num_adr, GFP_KERNEL);
+	if (!codec)
+		return -ENOMEM;
+
+	/*
+	 * Spec definition
+	 *   Register		Bit	Contents
+	 *   DevId_0 [7:4]	47:44	sdw_version
+	 *   DevId_0 [3:0]	43:40	unique_id
+	 *   DevId_1		39:32	mfg_id [15:8]
+	 *   DevId_2		31:24	mfg_id [7:0]
+	 *   DevId_3		23:16	part_id [15:8]
+	 *   DevId_4		15:08	part_id [7:0]
+	 *   DevId_5		07:00	class_id
+	 */
+	link_id = (link->adr[0] >> 48) & GENMASK(15, 0);
+	mfg_id = (link->adr[0] >> 24) & GENMASK(15, 0);
+	part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+	class_id = link->adr[0] & GENMASK(7, 0);
+
+	if (link->num_adr == 1) {
+		codec[0].name = devm_kasprintf(dev, GFP_KERNEL,
+					       "sdw:%x:%x:%x:%x", link_id,
+					       mfg_id, part_id, class_id);
+
+		if (!codec[0].name)
+			return -ENOMEM;
+	} else {
+		for (i = 0; i < link->num_adr; i++) {
+			unique_id = (link->adr[i] >> 40) & GENMASK(3, 0);
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id, unique_id);
```

Comment:
We will ignore the unique_id if there is no the same codec on a single link. For example, the unique is will be ignore if one rt1308 and one rt1011 on the same link. We don't have this case for now, but we may need to support it in the future.

Reply:
@bardliao  there are some complex patterns: AA,   ABA, AABBCC, ABCCAB...
The simple method is to set unique_id to 0XFF for single codec in _ADR table in acpi mach file, then it is easy to process it here.

Reply:
ok, I figured out  a simple method process different cases.

Reply:
> @bardliao there are some complex patterns: AA, ABA, AABBCC, ABCCAB...
> The simple method is to set unique_id to 0XFF for single codec in _ADR table in acpi mach file, then it is easy to process it here.

@RanderWang The issue is that the device name of a Slave with ignored unique id will not include the unique id. See [sdw_slave_add](https://github.com/thesofproject/linux/blob/upstream/soundwire/drivers/soundwire/slave.c#L38). So we won't find such codecs with a name with unique id. Yes, it might be a good idea to set unique_id to 0XFF in the acpi match table and we can have different codec name according to that.

Reply:
> ok, I figured out a simple method process different cases.

Great :)